### PR TITLE
Add support for tag-based environment deployment branch policy

### LINKED ISSUE #1974: [FEAT]: Add support for tag pattern to `github_repository_environment_deployment_policy` resource
### Describe the need

The resource `github_repository_environment_deployment_policy` has `branch_pattern` to add branch based protection so could you also implement the same for tag based pattern. The github API already supports this as `type=tag`.

https://docs.github.com/en/enterprise-cloud@latest/rest/deployments/branch-policies?apiVersion=2022-11-28#create-a-deployment-branch-policy

### SDK Version

_No response_

### API Version

_No response_

### Relevant log output

_No response_

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### PR DESCRIPTION
**Pull request by mcevoypeter**
Resolves #1974

----

### Before the change?

* The `github_repository_environment_deployment_policy` only supported branch-based policies even though the [GitHub API](https://docs.github.com/en/rest/deployments/branch-policies?apiVersion=2022-11-28#create-a-deployment-branch-policy) has support for both branch-based and tag-based policies.

### After the change?

* A `tag_pattern` attribute has been added to the `github_repository_environment_deployment_policy` resource.

### Pull request checklist
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been reviewed and added / updated if needed (for bug fixes / features)

### Does this introduce a breaking change?

Please see our docs on [breaking changes](https://github.com/octokit/.github/blob/master/community/breaking_changes.md) to help!

- [ ] Yes
- [x] No

----

### CODE REVIEW

**Comment by mcevoypeter** [comment]
I switched from a `pattern` + `type` approach to a `branch_pattern` + `tag_pattern` approach. The former was a breaking change, whereas the latter is not and is also simpler for users of the provider to use correctly.

**Comment by mcevoypeter** [comment]
@kfcampbell, is there anything I can do to help move this along?

File: website/docs/r/repository_environment_deployment_policy.html.markdown [hunk 1]
```diff
@@ -34,12 +36,45 @@ resource "github_repository_environment" "test" {
 }
 
 resource "github_repository_environment_deployment_policy" "test" {
-  repository 	   = github_repository.test.name
-  environment	   = github_repository_environment.test.environment
+  repository     = github_repository.test.name
+  environmen     = github_repository_environment.test.environment
```
Comment (brandocomando) [inline website/docs/r/repository_environment_deployment_policy.html.markdown:40]: typo environment

**Comment by brandocomando** [comment]
Does repository need to be updated as well? github_repository_environment
Currently it has 
deployment_branch_policy with custom_branch_policies.

Does it also need custom_tag_policies boolean?

**Comment by mcevoypeter** [comment]
> Does it \[the `github_repository_environment` resource \]also need custom_tag_policies boolean?

I don't think so. The [GitHub docs](https://docs.github.com/en/rest/deployments/environments?apiVersion=2022-11-28#create-or-update-an-environment--parameters) only lists two fields within the `deployment_branch_policy` object: `protected_branches` and `custom_branch_policies`.

**Comment by kfcampbell** [comment]
@mcevoypeter Sorry about the delay; it's been slow with holidays. Thank you for the contribution!

When running the newly-added tests, I get the following error that fails them: 

```
    testing.go:705: Step 0 error: After applying this step, the plan was not empty:
        
        DIFF:
        
        UPDATE: github_repository_environment_deployment_policy.test
          branch_pattern: "v*" => ""
          environment:    "environment/test" => "environment/test"
          id:             "tf-acc-test-7vk43:environment%2Ftest:13663474" => "tf-acc-test-7vk43:environment%2Ftest:13663474"
          repository:     "tf-acc-test-7vk43" => "tf-acc-test-7vk43"
          tag_pattern:    "v*" => "v*"
        
        
        
        STATE:
        
        (left out for now)
```

Can you reproduce this? Do they succeed for you?

**Comment by boredland** [comment]
Hi there! We're currently a bit blocked by this missing. Anything we could help with moving this forward?

**Comment by kfcampbell** [comment]
@boredland you're certainly welcome to take this code as it is, resolve the issues/conflicts, and submit as another PR! I'd be happy to review it should you decide to do so.

**Comment by mcevoypeter** [comment]
@boredland, go for it. I've been swamped since the beginning of January and can't get to this for the forseeable future. Let me know if there's anything I can do to help with hand-off.

**Comment by boredland** [comment]
Didn't find any time yet :(

**Comment by sumnerwarren** [comment]
@kfcampbell I took a stab at reworking this in #2165.

**Comment by nickfloyd** [comment]
Thanks for the efforts here; closing in favor of https://github.com/integrations/terraform-provider-github/pull/2165.

**Pull request closed without merging**